File: http-servers/typescript/bun-honojs/src/routes/params.ts
import { Hono } from "hono";
import { getCookie, setCookie } from "hono/cookie";
import { DEFAULT_LIMIT, MAX_FILE_BYTES, NULL_BYTE } from "../consts/defaults";
import {
  FILE_NOT_FOUND,
  FILE_NOT_TEXT,
  FILE_SIZE_EXCEEDS,
  INVALID_FORM_DATA,
  INVALID_JSON_BODY,
  INVALID_MULTIPART,
  makeError,
  ONLY_TEXT_PLAIN
} from "../consts/errors";

export const paramsRoutes = new Hono();

paramsRoutes.get("/search", (c) => {
  const q = c.req.query("q")?.trim() || "none";

  const limitStr = c.req.query("limit");
  const limitNum = Number(limitStr);
  const limit = Number.isSafeInteger(limitNum) ? limitNum : DEFAULT_LIMIT;

  return c.json({ search: q, limit });
});

paramsRoutes.get("/url/:dynamic", (c) => {
  const dynamic = c.req.param("dynamic");
  return c.json({ dynamic });
});

paramsRoutes.get("/header", (c) => {
  const header = c.req.header("X-Custom-Header")?.trim() || "none";
  return c.json({ header });
});

paramsRoutes.post("/body", async (c) => {
  let body: Record<string, unknown>;
  try {
    body = await c.req.json();
  } catch (err) {
    return c.json(makeError(INVALID_JSON_BODY, err), 400);
  }

  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    return c.json(makeError(INVALID_JSON_BODY, "expected a JSON object"), 400);
  }

  return c.json({ body });
});

paramsRoutes.get("/cookie", (c) => {
  const cookie = getCookie(c, "foo")?.trim() || "none";
  setCookie(c, "bar", "12345", { maxAge: 10, httpOnly: true, path: "/" });
  return c.json({ cookie });
});

paramsRoutes.post("/form", async (c) => {
  const contentType = c.req.header("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("application/x-www-form-urlencoded") && !contentType.startsWith("multipart/form-data")) {
    return c.json(
      makeError(INVALID_FORM_DATA, "expected content-type: application/x-www-form-urlencoded or multipart/form-data"),
      400
    );
  }

  let form: Record<string, string | File>;
  try {
    form = await c.req.parseBody();
  } catch (err) {
    return c.json(makeError(INVALID_FORM_DATA, err), 400);
  }

  if (typeof form !== "object" || form === null || Array.isArray(form)) {
    return c.json(makeError(INVALID_FORM_DATA, "expected form fields"), 400);
  }

  const name = typeof form.name === "string" && form.name.trim() !== "" ? form.name.trim() : "none";

  const ageStr = typeof form.age === "string" && form.age.trim() !== "" ? form.age.trim() : "0";
  const ageNum = Number(ageStr);
  const age = Number.isSafeInteger(ageNum) ? ageNum : 0;

  return c.json({ name, age });
});

paramsRoutes.post("/file", async (c) => {
  const contentType = c.req.header("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("multipart/form-data")) {
    return c.json(makeError(INVALID_MULTIPART, "expected content-type: multipart/form-data"), 400);
  }

  let form: FormData;
  try {
    form = await c.req.formData();
  } catch (err) {
    return c.json(makeError(INVALID_MULTIPART, err), 400);
  }

  const file = form.get("file");
  if (!(file instanceof File)) {
    return c.json(makeError(FILE_NOT_FOUND, "no file field named 'file' in form data"), 400);
  }
  if (!file.type || !file.type.startsWith("text/plain")) {
    return c.json(makeError(ONLY_TEXT_PLAIN, `received mimetype: ${file.type || "unknown"}`), 415);
  }
  if (file.size > MAX_FILE_BYTES) {
    return c.json(makeError(FILE_SIZE_EXCEEDS, `file size ${file.size} exceeds limit ${MAX_FILE_BYTES}`), 413);
  }

  const buffer = await file.arrayBuffer();
  const data = new Uint8Array(buffer);
  if (data.length > MAX_FILE_BYTES) {
    return c.json(makeError(FILE_SIZE_EXCEEDS, `file size ${data.length} exceeds limit ${MAX_FILE_BYTES}`), 413);
  }

  if (data.includes(NULL_BYTE)) {
    return c.json(makeError(FILE_NOT_TEXT, "file contains null bytes"), 415);
  }

  let content: string;
  try {
    content = new TextDecoder("utf-8", { fatal: true }).decode(data);
  } catch {
    return c.json(makeError(FILE_NOT_TEXT, "file is not valid UTF-8"), 415);
  }

  return c.json({
    filename: file.name,
    size: data.length,
    content
  });
});

File: servers/py-flask/src/blueprints/params.py
from __future__ import annotations

import json
from typing import Any

from flask import Blueprint, jsonify, request
from flask.typing import ResponseReturnValue

from bench_shared.consts import DEFAULT_LIMIT, MAX_FILE_BYTES, NULL_BYTE, SAFE_INT_LIMIT, SNIFF_LEN
from bench_shared.errors import (
    EXPECTED_FORM_CONTENT_TYPE,
    EXPECTED_MULTIPART_CONTENT_TYPE,
    FILE_NOT_FOUND,
    FILE_NOT_TEXT,
    FILE_SIZE_EXCEEDS,
    INVALID_FORM_DATA,
    INVALID_JSON_BODY,
    INVALID_MULTIPART,
    ONLY_TEXT_PLAIN,
)
from src.responses import json_error

bp = Blueprint("params", __name__)

_FORM_CONTENT_TYPES = ("application/x-www-form-urlencoded", "multipart/form-data")


def _strip_or(value: str | None, default: str) -> str:
    stripped = value.strip() if value else ""
    return stripped or default


def _parse_safe_int(value: str | None, default: int) -> int:
    if value is None or "." in value:
        return default
    try:
        num = int(value)
    except ValueError:
        return default
    if -SAFE_INT_LIMIT <= num <= SAFE_INT_LIMIT:
        return num
    return default


@bp.get("/search")
def search_params() -> ResponseReturnValue:
    return jsonify(
        {
            "search": _strip_or(request.args.get("q"), "none"),
            "limit": _parse_safe_int(request.args.get("limit"), DEFAULT_LIMIT),
        }
    )


@bp.get("/url/<dynamic>")
def url_params(dynamic: str) -> ResponseReturnValue:
    return jsonify({"dynamic": dynamic})


@bp.get("/header")
def header_params() -> ResponseReturnValue:
    return jsonify({"header": _strip_or(request.headers.get("X-Custom-Header"), "none")})


@bp.post("/body")
def body_params() -> ResponseReturnValue:
    try:
        payload: Any = json.loads(request.get_data())
    except ValueError as e:
        return json_error(INVALID_JSON_BODY, 400, e)
    if not isinstance(payload, dict):
        return json_error(INVALID_JSON_BODY, 400, "expected a JSON object")
    return jsonify({"body": payload})


@bp.get("/cookie")
def cookie_params() -> ResponseReturnValue:
    response = jsonify({"cookie": _strip_or(request.cookies.get("foo"), "none")})
    response.set_cookie("bar", "12345", max_age=10, httponly=True, path="/")
    return response


@bp.post("/form")
def form_params() -> ResponseReturnValue:
    if request.mimetype not in _FORM_CONTENT_TYPES:
        return json_error(INVALID_FORM_DATA, 400, EXPECTED_FORM_CONTENT_TYPE)

    name = _strip_or(request.form.get("name"), "none")
    age_raw = request.form.get("age")
    age = _parse_safe_int(age_raw if age_raw and age_raw.strip() else None, 0)
    return jsonify({"name": name, "age": age})


@bp.post("/file")
def file_params() -> ResponseReturnValue:
    if request.mimetype != "multipart/form-data":
        return json_error(INVALID_MULTIPART, 400, EXPECTED_MULTIPART_CONTENT_TYPE)

    upload = request.files.get("file")
    if upload is None:
        return json_error(FILE_NOT_FOUND, 400, "no file field named 'file' in form data")

    part_type = (upload.mimetype or "").lower()
    if not part_type.startswith("text/plain"):
        return json_error(ONLY_TEXT_PLAIN, 415, f"received mimetype: {upload.mimetype or 'unknown'}")

    data = upload.read()
    if len(data) > MAX_FILE_BYTES:
        return json_error(FILE_SIZE_EXCEEDS, 413, f"file size {len(data)} exceeds limit {MAX_FILE_BYTES}")

    if NULL_BYTE in data[:SNIFF_LEN]:
        return json_error(FILE_NOT_TEXT, 415, "file contains null bytes in header")
    if NULL_BYTE in data:
        return json_error(FILE_NOT_TEXT, 415, "file contains null bytes")

    try:
        content = data.decode("utf-8")
    except UnicodeDecodeError as e:
        return json_error(FILE_NOT_TEXT, 415, e)

    return jsonify({"filename": upload.filename, "size": len(data), "content": content})
